ludwig/api.py

```diff
@@ -546,6 +547,45 @@ def train(
 
         return train_stats
 
+    def kfold_cross_validate(
```

Comment:
As this function doesn't really use self (it only uses it to extract the model definition, but then it doesn't really touch self at all), I would prefer not to have it as a class method. It could be a static method, but even better it could be just a module function with the same parameters of `kfold_cross_validate` that just calls it and returns its outputs.
One would import it as `from ludwig.api import kfold_cross_validate`

Reply:
@w4nderlust   After reading your  comments, I understand you want me to do following:
* convert from class method to just a module function (`kfold_cross_validate()`) in the `api.py` module.
* for the`kfold_cross_validate()` module function, you would like it's signature to support all data formats: 'data_csv', 'data_df` and 'data_hdf5`
* Look to see if it is feasible to reuse the `experiment` functionality instead of the `train` and `predict` functions from the other modules.

Unless I hear otherwise, I'll start implementing the above changes.

Reply:
Yes, this is correct. The only caveat is about data formats. I believe hdf5+json would be kinda difficult to make work because of the splitting, so I would say `csv`, `df` and `dict` would be sufficient.

Reply:
I've looked at the feasibility of implementing this comment:

> in the inner loop of the kfold you use full_train and then predict https://github.com/uber/ludwig/pull/627/files#diff-ecb29d746617ca1156fc28dfb325713cL371-L386, is there a way to use experiment instead? 

In the following code fragments: `train_fold` and `test_fold` are pandas data frames based on current design when we converted from temporary csv files to pandas data frames.

Psuedo-code for current design for the k-fold cv:
``` 
def kfold_cross_validate():
    for train_fold, test_fold in data_folds:
        full_train(train_fold)
        predict(test_fold
```

Psuedo code desired design:
``` 
def kfold_cross_validate():
    for train_fold, test_fold in data_folds:
        experiment(train_fold, test_fold)
```

To support this change, I believe these modifications are needed:

#### Change signature for `experiment()` function
If we keep the current design where `train_fold` and `test_fold` pandas dataframes, I'll need to change the signature for `experiment()` to accept pandas dataframes.  The current signature for `experiment()` is this
https://github.com/uber/ludwig/blob/c167b33079e32983196f0c8a75ce486883180164/ludwig/experiment.py#L47-L79

At minimum I have to add these parameters to `experiment()`'s signature:
* `data_train_df`
* `data_test_df`

For completeness sake, do we want to add these as well?
* `data_validation_df`
* `data_df`

#### Refactor `experiment()` into two functions
Currently `experiment()` returns a string indicating the file path for the experiment's results file.  With this design, we'll have to do k-fold reads from the files.  To avoid these reads from files, I'll need to break out this section of `experiment()` (https://github.com/uber/ludwig/blob/c167b33079e32983196f0c8a75ce486883180164/ludwig/experiment.py#L188-L248) to a separate function, something called like `run_an_experiment()`.  The return from this new function will be the `test_results`.

With this change the refactored `experiment()`  looks something like this
``` 
def run_an_experiment():
    # extracted from orginal experiment() function
    # https://github.com/uber/ludwig/blob/c167b33079e32983196f0c8a75ce486883180164/ludwig/experiment.py#L188-L248

    return test_results
    
def experiment():
    test_results = run_an_experiment()

    # the remaining part of this function to create the 
    # experiment result file from test_results
    # https://github.com/uber/ludwig/blob/c167b33079e32983196f0c8a75ce486883180164/ludwig/experiment.py#L250-L286

    return experiment_dir_name

```

The kfold function would now look like this:
``` 
def kfold_cross_validate():
    for train_fold, test_fold in data_folds:
        test_results = run_an_experiment(train_fold, test_fold)
        
        # extract statistics for each fold

    # consolidate all the fold test results to build the 
    # kfold cv results
```

With this change the return for `experiment()` stays the same.  So existing code should not break.  At the same time, we can get access to the in-memory data structures from `run_an_experiment()` to be used in k-fold CV function.

Do these changes seem reasonable?

Reply:
Yes it makes sense! we can follow the same naming pattern for `train` and `full_train` meaning the outpost function is `full_experiment` and the inmost function is `experiment`. But before doing it i would first refactor all usages of the current `experiment` to `full_experiment` and then later add the new `experiment` function that is your `run_a_experiment`.
Also the new `experiment` inmost function should support all parameters of `full_train` and `predict` even if in the kfold case we will not use them because `full_experiment` will.
Does this make sense?

Reply:
yep...I'll get started tonight.

Reply:
@w4nderlust   Just an update.  As of this posting, I've completed the following:
* Moved `kfold_cross_validate()` from LudwigModel to `experiment.py` module
* refactored `experiment.py` to implement `full_experiment()` and `experiment()` functions
* add support for pandas dataframes: data_df, data_train_df, data_validation_df, data_test_df

I'm still in process of adapting the kfold_cv function to use new `experiment()` implementation.

Reply:
Thank you so much @jimthompson5802 ! One little thing I was not precise about: when I was saying to make `kfold_cross_validate` a module function instead of a class function of `LudwigModel` I intended a module function of the `api.py` module, not the `experiment.py` module. Reason being that, although all modules are exposed anyway, the ones in API are the only one documented and the ones people who don't want to adventure in the inner workings of the Ludwig code should use.

Reply:
NP...my misunderstanding.  I'll move `kfold_cross_validate()` back to `api.py` module as a module function.  Not a big deal.

Reply:
@w4nderlust 
### Summary of changes:
* Implemented `kfold_cross_validate()` as module function in `api.py`.  Note: To solve a circular import error, all the k-fold cross validation functionality are in `experiment.py`.  This supports both the cli and api feature.  `api.kfold_cross_validate()` is a wrapper that calls the core computations in `experiment.py`

#### pytest log from local testing:
``` 
============================= test session starts ==============================
platform linux -- Python 3.6.8, pytest-5.3.5, py-1.8.1, pluggy-0.13.1
rootdir: /opt/project
collected 81 items

tests/integration_tests/test_api.py .                                    [  1%]
tests/integration_tests/test_contrib_wandb.py .                          [  2%]
tests/integration_tests/test_experiment.py ...................           [ 25%]
tests/integration_tests/test_kfold_cv.py ...                             [ 29%]
tests/integration_tests/test_server.py .                                 [ 30%]
tests/integration_tests/test_visualization.py .......................... [ 62%]
                                                                         [ 62%]
tests/integration_tests/test_visualization_api.py ...................... [ 90%]
                                                                         [ 90%]
tests/ludwig/models/modules/test_encoder.py ....                         [ 95%]
tests/ludwig/utils/test_data_utils.py .                                  [ 96%]
tests/ludwig/utils/test_image_utils.py ..                                [ 98%]
tests/ludwig/utils/test_normalization.py .                               [100%]

=============================== warnings summary ===============================
<<<<<<<<<< Removed warning output >>>>>>>>>>>>>>
================ 81 passed, 7493 warnings in 377.42s (0:06:17) =================
```

#### Sample `kfold_cross_validate()` api code:
``` 
import shutil
import pprint
import yaml

from ludwig.api import kfold_cross_validate

# clean out prior results
shutil.rmtree('./results', ignore_errors=True)

# execute k-fold cross validation run
kfold_training_stats, kfold_split_indices = kfold_cross_validate(
    3,
    model_definition_file='./model1_definition.yaml',
    data_csv='./data/train.csv',
)


print('\nk-fold cross validation statistics from file:')
pprint.pprint(kfold_training_stats['overall'])



starting 3-fold cross validation
<<<<<< delete training related output >>>>>>>>>
completed 3-fold cross validation

k-fold cross validation statistics from file:
{'Survived': {'accuracy_mean': 0.41526374859708187,
              'accuracy_std': 0.01657106965233829,
              'average_precision_macro_mean': 0.5461645855677718,
              'average_precision_macro_std': 0.049048669150776075,
              'average_precision_micro_mean': 0.5461645855677718,
              'average_precision_micro_std': 0.049048669150776075,
              'average_precision_samples_mean': 0.5461645855677718,
              'average_precision_samples_std': 0.049048669150776075,
              'loss_mean': 5.858693320876955,
              'loss_std': 0.4936412256915484,
              'roc_auc_macro_mean': 0.6901250280742485,
              'roc_auc_macro_std': 0.03752298746825169,
              'roc_auc_micro_mean': 0.6901250280742485,
              'roc_auc_micro_std': 0.03752298746825169},
 'combined': {'accuracy_mean': 0.41526374859708187,
              'accuracy_std': 0.01657106965233829,
              'loss_mean': 5.858693320876955,
              'loss_std': 0.4936412256915484}}

```

#### Example cli usage with results files:
``` 
ludwig experiment \
  --model_definition_file sandbox/model1_definition.yaml \
  --data_csv sandbox/data/train.csv \
  --output_directory sandbox/results \
  -kf 4


ls -l sandbox/results
total 228
-rw-r--r-- 1 root root  60542 Feb  7 02:22 kfold_split_indices.json
-rw-r--r-- 1 root root 172029 Feb  7 02:22 kfold_training_statistics.json
```

I believe this addresses all the comments.  I'm marking this resolved.

Let me know if any other revisions are needed.

Reply:
Looks good, thanks, let me do another pass of review before merging.